local_planner/src/nodes/star_planner.cpp

```diff
@@ -157,90 +141,66 @@ void StarPlanner::buildLookAheadTree() {
     Eigen::Vector3f origin_position = tree_[origin].getPosition();
     int old_origin = tree_[origin].origin_;
     Eigen::Vector3f origin_origin_position = tree_[old_origin].getPosition();
-
-    bool node_valid = true;
-
-    // crop pointcloud
-    pcl::PointCloud<pcl::PointXYZ> cropped_cloud;
-    Eigen::Vector3f closest_point;  // unused
-    bool hist_is_empty = false;     // unused
-    int backoff_points_counter = 0;
-    double distance_to_closest_point;
-    histogram_box_.setBoxLimits(toPoint(origin_position), ground_distance_);
-
-    filterPointCloud(cropped_cloud, closest_point, distance_to_closest_point,
-                     backoff_points_counter, complete_cloud_, min_cloud_size_,
-                     min_dist_backoff_, histogram_box_, origin_position,
-                     min_realsense_dist_);
-
-    if (origin != 0 && backoff_points_counter > 20 &&
-        cropped_cloud.points.size() > 160) {
+    bool hist_is_empty = false;  // unused
+
+    // build new histogram
+    std::vector<int> z_FOV_idx;
+    int e_FOV_min, e_FOV_max;
+    calculateFOV(h_FOV_, v_FOV_, z_FOV_idx, e_FOV_min, e_FOV_max,
+                 tree_[origin].yaw_,
+                 0.0);  // assume pitch is zero at every node
+
+    Histogram propagated_histogram = Histogram(2 * ALPHA_RES);
+    Histogram histogram = Histogram(ALPHA_RES);
+
+    propagateHistogram(propagated_histogram, reprojected_points_,
```

Comment:
If I understand correctly `pose_` and `pointcloud_` are not getting updated while being inside the loop ` while (n < n_expanded_nodes_) ` so all the histogram calculations could be done outside the loop not to waste compute.

Reply:
Biiig catch!! But it is the other way round than you thought. The position should of course change. We want to evaluate the possible movement directions at the location of every tree node and not constantly evaluate the current position.